Add integration test infrastructure

hargut opened the pull request:
Hello,

this PR contains a first TCPRoute (#336) implementation including an end-to-end integration test.

The test infrastructure was based around:
- command executions `kind`, `kubectl` and `podman`
- executions are abstracted through structs like `KindCluster`, ` Image`, `Container` and `KustomizeDeployments`

The integration test can be executed with:
`cargo test --test integration_tcp_route --profile test`

The test does not install any of the required commands.

I've re-factored/consolidated existing files and tried to find a structure that is suitable, but I'm not sure if the result is the best approach for all the requirements in this project.

Along within the PR are structured error handling and some documentation/overviews.

Many thanks in advance and have a great time. :sunflower: 

Kind regards,
Harald

shaneutt reviewed (commented):
Thank you for putting in this PR! I took a first pass here, and I'm noticing a lot of files have been changed. What are your thoughts on reducing this down to _just_ the tcproute controller and the test suite? Just trying to see if we can make this a little smaller (and then other changes which are more like refactors, etc can be separate PRs). LMKWYT?

hargut committed:
fix(config): add metallb resources

Signed-off-by: Harald Gutmann <harald@gutmann.one>

hargut committed:
fix(config): default deployment and pod startup issues

add controlplane health check
expose dataplane service port
remove missing patch file

Signed-off-by: Harald Gutmann <harald@gutmann.one>

hargut committed:
feat(tests): add integration test infrastructure

add integreation test for cluster setup
build images in develop and release mode
deploy cluster with metallb and default config
check image tag updates

Signed-off-by: Harald Gutmann <harald@gutmann.one>

shaneutt commented:
(just a note that I will get back to reviewing, it just might be a couple of days!)

shaneutt commented on build/Containerfile.controlplane:
```diff
@@ -18,6 +18,8 @@ COPY controlplane/ controlplane/
 
 COPY dataplane/ dataplane/
 
+COPY tests/ tests/
+
```
This should not be required in the container image right?

shaneutt replied:
Ok. There's probably other ways we can manage this (having a different Cargo workspace for builds) but this will be quicker for now, so let's keep it for now. :+1:

shaneutt commented on build/Containerfile.udp_test_server:
```diff
@@ -20,6 +18,8 @@ COPY controlplane/ controlplane/
 
 COPY dataplane/ dataplane/
 
+COPY tests/ tests/
+
```
This should not be required in the container image right?

shaneutt reviewed (changes requested):
I had some time this morning and took a first pass, thanks again for taking the time to create this!

Most of my initial comments are questions about some specific choices, and some minor suggested changes. LMKWYT, and then in my next pass I'll dig deeper into the test code.

Very much appreciate you trimming this down to focus on the test suite, thank you. :)

hargut commented on build/develop/Containerfile.controlplane:
```diff
@@ -0,0 +1,19 @@
+# ------------------------------------------------------------------------------
+# Image
+# ------------------------------------------------------------------------------
+
+FROM debian:trixie-slim
+
+LABEL org.opencontainers.image.source=https://github.com/kubernetes-sigs/blixt
+
+WORKDIR /
+
+ARG BUILD_TIMESTAMP
+RUN echo $BUILD_TIMESTAMP && cp /workspace_mount/target/debug/controller /controller \
+    && /usr/bin/ldd /controller
+
+USER 1000:1000
+
+EXPOSE 8081
+
+ENTRYPOINT [ "/controller" ]
```
The develop images had been intended to improve the round-trip-time for deployments to local kind clusters. Using a volume mount and the already compiled binary from the standard `cargo build` brings container build times down to approx. 15 secs per image. With some rollout-restart times added, the deployments are live within under a minute. Removing the build context from the podman build command would even further drop the build times for these images to approx. 2-3 secs per image.

Using the  existing images a full project recompile is required to get the containers ready for deployment. Once a file in the `COPY` sources is changed, the layers built on top have to be fully rebuilt (including dependency downloads). The image build times for this cycle are on my machine approx. 3-5 mins per image.

There are no insufficiencies with the existing images except that the round-trip-times are comparably long. Considering that often two images will need to be rebuilt the developer experience is not ideal when using these images for local deployments during developing.


The develop container images are not needed and could be removed.

hargut replied:
Removed.

hargut commented on config/manager/manager.yaml:
```diff
@@ -47,14 +47,12 @@ spec:
-            path: /healthz
+          grpc:
             port: 8081
           initialDelaySeconds: 15
           periodSeconds: 20
         readinessProbe:
-          httpGet:
-            path: /readyz
+          grpc:
```
Yes, the rust dataplane manager did not provide any probe endpoints. The pod had not been starting up correctly.

> Liveness probe failed: Get "http://10.244.0.14:8081/healthz": dial tcp 10.244.0.14:8081: connect: connection refused
> Readiness probe failed: Get "http://10.244.0.14:8081/readyz": dial tcp 10.244.0.14:8081: connect: connection refused

hargut commented on tests/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+license.workspace = true
+repository.workspace = true
+version.workspace = true
+
```
That's for sure a good idea. :+1:

hargut replied:
Added.

shaneutt reviewed (changes requested):
I took another pass, this time deeper into the test code. I noticed that we're doing a lot of trying to model and interact with the system, the container runtime, and build images.

I think for this first iteration our workflow needs to be something lighter, such as:

1. use `kind` to create a cluster
2. patch in metallb to enable `LoadBalancer` type `Services`
  a. as a hack for now, I think shelling out to `kubectl kustomize` for that is OK
4. do some verification of the blixt controlplane and dataplane (using the Kubernetes API) to make sure everything's working
5. run our tests
6. teardown

Limiting how much we're doing shell calls, etc. LMKWYT? :thinking:

hargut commented on tests-integration/tests/defaults.rs:
```diff
@@ -0,0 +1,132 @@
+    images.load_images(cluster).await?;
+
+    let gateway_api_metallb = vec![
+        "https://github.com/metallb/metallb/config/native?ref=v0.13.11",
```
You're correct it is the version I followed through the existing go based integration tests and its dependencies.

At the point in time, after trying to troubleshoot and figure out why I was failing with cloud-provider-kind to get any TCP packets onto the dataplane pods I chose to exactly reproduce a "known to be working" state. That's why I followed the go dependencies and the versions used in there, including this metallb version.

There was no AI involved in the process and it was fully intentional choice to stick with exactly the setup used in the existing integration tests.

The version can be updated, no issues from my side with that.

hargut commented on tests-integration/tests/defaults.rs:
```diff
@@ -0,0 +1,132 @@
+        )
+        .await?;
+
+    let deployments = vec!["../config/metallb/", "../config/default/"];
```
Will move these files.

hargut commented on tests-integration/tests/integration_cluster_images.rs:
```diff
@@ -0,0 +1,78 @@
+    assert!(controlplane_deployment.is_ok());
+
+    // check if image tag on deployment is correct
+    let deployment = controlplane_deployment.unwrap();
+    if let Some(spec) = deployment.spec {
+        if let Some(pod) = spec.template.spec {
+            assert!(
+                pod.containers
+                    .iter()
+                    .any(|c| if let Some(image) = &c.image {
+                        return image.ends_with("blixt-controlplane:integration-release");
+                    } else {
+                        false
+                    })
+            )
+        } else {
+            assert!(false)
+        };
+    } else {
+        assert!(false)
+    };
```
When deploying the default, without invoking kubectl to set the image to the correct tags the deployment contains `latest` as tag and the pods do not come up correctly.
As the cluster setup and the image build was done in parallel using `try_join` this check is done at a later point in time, after the images are loaded into the cluster the workloads are updated with the correct image tag and the rollout waits for healthy pods before continuing.

hargut commented on tests-integration/src/infrastructure/mod.rs:
```diff
@@ -0,0 +1,278 @@
+        self
+    }
+
+    /// set the working directory for the commands execution
+    pub fn working_dir<P: AsRef<Path>>(mut self, dir: P) -> Self {
+        self.cmd.current_dir(dir);
+        self
+    }
+
+    async fn run(&mut self) -> Result<(), AsyncCommandError> {
+        info!("run: {:?}", self.cmd);
+        let exit_status = self
+            .cmd
+            .spawn()
+            .map_err(AsyncCommandError::Spawn)?
+            .wait()
+            .await
+            .map_err(AsyncCommandError::Wait)?;
+
+        if !exit_status.success() {
+            return Err(AsyncCommandError::ExitStatus(exit_status.code()));
+        }
+
+        Ok(())
+    }
+
+    async fn read(&mut self) -> Result<(String, String), AsyncCommandError> {
+        info!("read: {:?}", self.cmd);
+        let output = self.cmd.output().await.map_err(AsyncCommandError::Output)?;
+
+        if !output.status.success() {
+            return Err(AsyncCommandError::ExitStatus(output.status.code()));
+        }
+
+        let stdout = String::from_utf8_lossy(output.stdout.as_slice()).to_string();
+        let stderr = String::from_utf8_lossy(output.stderr.as_slice()).to_string();
+
+        Ok((stdout, stderr))
+    }
+}
```
Ok.

hargut replied:
Sorry, but I think this is not a good idea. This would require to duplicate exit status code checks for any other command invocation. I'll make it private for now, as with the images the only public usage is removed.

shaneutt replied:
Alright, I'll consider this resolved for now and take another pass.

hargut commented on tests-integration/tests/defaults.rs:
```diff
@@ -0,0 +1,132 @@
+/*
```
Ok.

hargut replied:
Removed.

hargut commented on tests-integration/tests/defaults.rs:
```diff
@@ -0,0 +1,132 @@
+    Ok(images)
+}
+
+pub(crate) fn containers(cargo_workspace_dir: &str) -> Result<Vec<Container>> {
```
Ok.

hargut commented on tests-integration/tests/integration_cluster_images.rs:
```diff
@@ -0,0 +1,78 @@
+use tracing_subscriber::EnvFilter;
+
+#[tokio::test]
+async fn integration_cluster_images() -> Result<()> {
```
Fine with me.

hargut replied:
Removed the tokio test, as this does not work with return values without implementing a Terminated trait.
Updated the function signature to return the KindCluster.

Added a tokio test that calls the function and returns a Result<()>.

hargut commented on tests-integration/tests/defaults.rs:
```diff
@@ -0,0 +1,132 @@
+
+    let gateway_api_metallb = vec![
+        "https://github.com/metallb/metallb/config/native?ref=v0.13.11",
+        "https://github.com/kubernetes-sigs/gateway-api/config/crd/experimental?ref=v1.2.1",
```
Updated.

hargut commented on tests-integration/tests/defaults.rs:
```diff
@@ -0,0 +1,132 @@
+        pre_build_commands: vec![],
+    };
+    Ok(images)
+}
```
Dropped.

shaneutt reviewed (approved):
Thanks for your patience @hargut. I'm sorry about the review delays, it's been a combination of extremely busy days and PTO the last couple weeks. Also I'm the only active maintainer at the moment (which I'll try to see about improving, but at least for the moment that means everything waits on me unfortunately).

In sensitivity of how long its taking for me to review, for this next pass of review I decided to focus on just making sure the basic framework is there and not worry about too much optimization of any kind.

As a side note for future contributions: try to break things out as much as is feasible into small chunked deliverables, as this will give us more overall velocity.

For now though, I think this is a solid starting point we can continue to iterate on as we see need:

/approve
/lgtm

Thank you very much for your contributions here, they are very much appreciated!

shaneutt commented:
Looks like it merged prior to builds and tests completing (something might be misconfigured to allow that, I'll take a look :thinking:), then failed on `main`.

We'll need a patch. I'm in favor of patching to `main` rather than reverting.
